hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/EmbeddedForeignKeyDescriptor.java

```diff
@@ -255,46 +260,49 @@ public Predicate generateJoinPredicate(
 		final String lhsTableExpression = lhs.getTableExpression();
 		if ( lhsTableExpression.equals( keyColumnContainingTable ) ) {
 			assert rhsTableExpression.equals( targetColumnContainingTable );
-			return getPredicate( lhs, rhs, creationContext, keyColumnExpressions, targetColumnExpressions );
+			return getPredicate( lhs, rhs, creationContext, keyColumnMappings, targetColumnMappings );
 		}
 		else {
 			assert rhsTableExpression.equals( keyColumnContainingTable );
-			return getPredicate( lhs, rhs, creationContext, targetColumnExpressions, keyColumnExpressions );
+			return getPredicate( lhs, rhs, creationContext, targetColumnMappings, keyColumnMappings );
 		}
 	}
 
 	private Predicate getPredicate(
 			TableReference lhs,
 			TableReference rhs,
 			SqlAstCreationContext creationContext,
-			List<String> lhsExpressions,
-			List<String> rhsColumnExpressions) {
+			ColumnMappings lhsMappings,
+			ColumnMappings rhsMappings) {
 		final Junction predicate = new Junction( Junction.Nature.CONJUNCTION );
-		for ( int i = 0; i < lhsExpressions.size(); i++ ) {
-			final JdbcMapping jdbcMapping = jdbcMappings.get( i );
-			final ComparisonPredicate comparisonPredicate = new ComparisonPredicate(
-					new ColumnReference(
-							lhs,
-							lhsExpressions.get( i ),
-							false,
-							null,
-							null,
-							jdbcMapping,
-							creationContext.getSessionFactory()
-					),
-					ComparisonOperator.EQUAL,
-					new ColumnReference(
-							rhs,
-							rhsColumnExpressions.get( i ),
-							false,
-							null,
-							null,
-							jdbcMapping,
-							creationContext.getSessionFactory()
-					)
-			);
-			predicate.add( comparisonPredicate );
-		}
+		lhsMappings.forEachColumn(
```

Comment:
The more I see it the more I really like your naming pattern here better - `#forEachColumn` as opposed to `#visitColumns`.  It better matches with the Java naming.  

I'm actually thinking of a general refactoring to make that change.  What do y'all think?